### PR TITLE
Issue #3429 - Adding WebAppProvider.scan() method to JMX

### PR DESCRIPTION
**Pull request by joakime**
+ Fixing up some bad JMX attributes as well.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentManager.java [hunk 1]
```diff
@@ -181,7 +180,7 @@ public void addAppProvider(AppProvider provider)
         if (isRunning())
             throw new IllegalStateException();
         _providers.add(provider);
-        addBean(provider);        
+        addManaged(provider);
```
Comment (sbordet) [inline jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentManager.java]: I think we want to leave `addBean()` here.
If user code passes in an already started `AppProvider` we should not manage it.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java [hunk 2]
```diff
@@ -321,4 +323,15 @@ public void setScanInterval(int scanInterval)
     {
         _scanInterval = scanInterval;
     }
+
+    /* ------------------------------------------------------------ */
```
Comment (sbordet) [inline jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java]: Remove this line!

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/jmx/WebAppProviderMBean.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.jetty.deploy.providers.WebAppProvider;
+import org.eclipse.jetty.server.handler.jmx.AbstractHandlerMBean;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+
+@ManagedObject("WebAppProvider mbean wrapper")
+public class WebAppProviderMBean extends AbstractHandlerMBean
```
Comment (sbordet) [inline jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/jmx/WebAppProviderMBean.java]: Why this class?
Can't the only JMX method in this class be moved to `WebAppProvider`?

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/jmx/WebAppProviderMBean.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        super(managedObject);
+    }
+
+    @ManagedAttribute("List of monitored resource directories")
+    public List<String> getMonitoredDirectories()
+    {
+        return ((WebAppProvider) _managed).getMonitoredResources().stream()
+                .map((r) -> r.getURI().toASCIIString())
+                .collect(Collectors.toList());
```
Comment (sbordet) [inline jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/jmx/WebAppProviderMBean.java]: We can monitor also file resources, right? If so, then it's not correct to call this method `getMonitoredDirectories()`.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentManager.java (hunk 1 above)
Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentManager.java]: Typically, the `addAppProvider` occurs in the `jetty-deploy.xml`, which makes it difficult to setup managed (or not).

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java (hunk 2 above)
Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java]: Just this one? or all of them in the entire file?

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/jmx/WebAppProviderMBean.java (hunk 3 above)
Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/jmx/WebAppProviderMBean.java]: That would make a different API.
JMX just wants to see the collection as a set of Strings, not a complex object that needs to be Serializable.
This is exactly the same reason jmx.DeploymentManagerMBean exists for as well.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/jmx/WebAppProviderMBean.java (hunk 4 above)
Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/jmx/WebAppProviderMBean.java]: Yup. good point. fixing it.

**Comment by joakime** [comment]
@sbordet applied your requested changes

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentManager.java (hunk 1 above)
Comment (sbordet) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentManager.java]: It's a `public` method people can call so you can get a `AppProvider` that is already started and used somewhere else and you don't want to manage.
In the normal case you mention, `addBean()` and `addManaged()` will behave the same. The former also covers my example use case, so I don't see why change it.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/jmx/DeploymentManagerMBean.java [hunk 5]
```diff
@@ -95,9 +95,10 @@ private String toRef(App app)
         return apps;
     }
 
-    public Collection<AppProvider> getAppProviders()
+    @ManagedAttribute("Registered AppProviders")
+    public List<String> getAppProviders()
     {
-        return _manager.getAppProviders();
+        return _manager.getAppProviders().stream().map((ap) -> ap.toString()).collect(Collectors.toList());
```
Comment (sbordet) [inline jetty-deploy/src/main/java/org/eclipse/jetty/deploy/jmx/DeploymentManagerMBean.java]: Just a nit, you can use `.map(String::valueOf)` here.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java [hunk 6]
```diff
@@ -310,15 +312,26 @@ public void setMonitoredDirectories(Collection<String> directories)
...
 
-    /* ------------------------------------------------------------ */
+
     public void setScanInterval(int scanInterval)
     {
         _scanInterval = scanInterval;
     }
+
+
+    @ManagedOperation("Scan the monitored directories")
```
Comment (sbordet) [inline jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java]: Add `impact="ACTION"` to `@ManagedOperation`.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/jmx/DeploymentManagerMBean.java (hunk 5 above)
Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/jmx/DeploymentManagerMBean.java]: done

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java (hunk 6 above)
Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java]: done.

**Pull request merged**